### DISCUSSION TITLE
Field "image" must not have a selection since type "String"

### DISCUSSION THREAD

**Discussion by DWboutin** [gatsbyjs/gatsby#37228] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I can't use `childImageSharp` in GraphQL queries to optimize images.

### Reproduction Link

https://github.com/DWboutin/gatsby-image-problem

### Steps to Reproduce

1. `npm install`
2. `npm start`
3. the error will appear
...

### Expected Result

Should return a gatsby image object instead of the string

### Actual Result

Return a string and throws an error

```
{
  "errors": [
    {
      "message": "Field \"image\" must not have a selection since type \"String\" has no subfields.",
      "locations": [
        {
          "line": 5,
          "column": 15
        }
      ]
    }
  ]
}
```

### Environment

```shell
System:
    OS: macOS 13.0.1
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Chrome: 108.0.5359.98
    Safari: 16.1
  npmPackages:
    gatsby: ^5.0.0 => 5.2.0 
    gatsby-image: ^3.11.0 => 3.11.0 
    gatsby-plugin-image: ^3.2.0 => 3.2.0 
    gatsby-plugin-mdx: ^5.1.0 => 5.2.0 
    gatsby-plugin-netlify-cms: ^7.0.0 => 7.2.0 
    gatsby-plugin-sharp: ^5.1.0 => 5.2.0 
    gatsby-plugin-styled-components: ^6.0.0 => 6.2.0 
    gatsby-plugin-typescript: ^5.0.0 => 5.2.0 
    gatsby-source-filesystem: ^5.0.0 => 5.2.0 
    gatsby-transformer-remark: ^6.0.0 => 6.2.0 
    gatsby-transformer-sharp: ^5.1.0 => 5.2.0 
  npmGlobalPackages:
    gatsby-cli: 5.0.0
```

### Config Flags

no flags

**Comment by DWboutin** [comment]
I already tried all (found) issues answers. Very sorry if it's a duplicate, but it's been days that I'm on that and I can't figure it out.

Thanks a lot for your help!

**Comment by LekoArts** [comment]
This post explains that you'll need to use `gatsby-remark-relative-images` with Netlify CMS: https://www.frontendstumbles.com/using-gatsby-image-with-netlify-cms/

**Reply by DWboutin** [reply]
A post from 3 years ago about a plugin who is not maintained. 

Do you have other solutions like changing the paths of my images or something related to Gatsby?

**Comment by vinassefranche** [comment]
We're having the same issue on our side. Any idea of where to look for to fix it?

Thanks a lot for creating this discussion and to anybody helping us!

**Comment by AndrewNow** [comment]
Encountering this issue as well, very frustrating :(

**Comment by vinassefranche** [comment]
Hi!

We fixed the issue on our side by using this `gatsby-plugin-relative-images` custom plugin in our project:

- gatsby-node.js:
  ```js
  const path = require("path");
  const deepMap = require("deep-map");
  const _ = require("lodash");

  const fileNodes = [];

  const fmImagesToRelative = node => {
    // Save file references
    fileNodes.push(node);
    // Only process markdown files
    if (node.internal.type === 'MarkdownRemark' || node.internal.type === 'Mdx') {
      // Convert paths in frontmatter to relative
      // eslint-disable-next-line no-inner-declarations
      function makeRelative(value) {
        if (_.isString(value) && path.isAbsolute(value)) {
          let imagePath;
          const foundImageNode = _.find(fileNodes, file => {
            if (!file.dir) return;
            imagePath = path.join(file.dir, path.basename(value));
            return path.normalize(file.absolutePath) === imagePath;
          });
          if (foundImageNode) {
            const nodeFilePath = path.join(node.internal.type === 'Mdx' ? node.internal.contentFilePath : node.absolutePath, '..')
            return path.relative(
              nodeFilePath,
              imagePath
            );
          }
        }
        return value;
      }
      // Deeply iterate through frontmatter data for absolute paths
      deepMap(node.frontmatter, makeRelative, { inPlace: true });
    }
  };

  exports.onCreateNode = ({ node }) => {
    fmImagesToRelative(node);
  }
  ```

- index.js:
  ```js
  //no-op
  ```

- package.json:
  ```json
  {
    "main": "index.js",
    "description": "largely copied from https://github.com/zslabs/gatsby-remark-relative-images-v2 with a fix for MDX"
  }
  ```
Just put this files in a `plugins/gatsby-plugin-relative-images/` folder in your project and make sure you have this in the plugins in your gatsby-config.js file:
```
{
  resolve: "gatsby-plugin-relative-images",
}
```

As said in the package.json it's largely copied from https://github.com/zslabs/gatsby-remark-relative-images-v2. We just added the support of `Mdx` in addition to `MarkdownRemark`.

I hope it helps until there's a real solution somewhere.
